quickinput-sample/src/promptCommandWithHistory.ts

```diff
@@ -0,0 +1,111 @@
+import * as fs from 'fs';
+import { window, Disposable } from 'vscode';
+import { QuickPickItem } from 'vscode';
+import * as _ from 'lodash';
+
+export const historyPath = `${process.env.HOME}/.vscode-cmd-history`;
+
+/**
+ * A command prompt with history
+ * 
+ */
+export async function promptCommand() {
+	const command = await pickCommand();
+	if (command) {
+		window.showInformationMessage(`You picked the following command: '${command}'`)
+	}
+}
+
+class CommandItem implements QuickPickItem {
+	public label: string;
+	public description?: string;
+	constructor(label: string, description?: string) {
+		this.label = label;
+		this.description = description;
+	}
+}
+class HistoryItem extends CommandItem {
+	constructor(label: string, description?: string) {
+		super(label, description);
+	}
+}
+class InputItem extends CommandItem {
+	constructor(public label: string) {
+		super(label, '(current input)');
+	};
+}
+
+async function pickCommand() {
+	const disposables: Disposable[] = [];
+	let commandsItems: CommandItem[] = [];
+	let currentValue: string | undefined = undefined;
+	let historyShouldBeUpdated = false;
+
+	try {
+		return await new Promise<string | undefined>((resolve, reject) => {
+			const input = window.createQuickPick<CommandItem>();
+			input.placeholder = 'Type a command';
+			input.items = commandsItems;
+
+			const updateQuickPick = (value?: string): void => {
+				if (!value) {
```

Comment:
We don't do any diffing behind the scenes (at least yet), it would be more efficient to only update the items when they changed.